compiler/src/iree/compiler/API2/Internal/Embed.cpp

```diff
@@ -205,20 +206,28 @@ Error *Source::openFile(const char *filePath) {
 }
 
 Error *Source::wrapBuffer(const char *bufferName, const char *buffer,
-                          size_t length) {
-  // Sharp edge: MemoryBuffer::getMemBuffer will peek one past the passed length
-  // to verify a nul terminator, but this makes the API really hard to ensure
-  // memory safety for. For our API, we just require that the buffer is nul
-  // terminated and that the nul is included in the length. We then subtract
-  // by 1 when constructing the underlying MemoryBuffer. This is quite sad :(
-  if (length == 0 || buffer[length - 1] != 0) {
-    return new Error("expected nul terminated buffer");
+                          size_t length, bool isNullTerminated) {
+  std::unique_ptr<llvm::MemoryBuffer> memoryBuffer;
+  if (isNullTerminated) {
+    // Sharp edge: MemoryBuffer::getMemBuffer will peek one past the passed
+    // length to verify a null terminator, but this makes the API really hard to
+    // ensure memory safety for. For our API, we just require that the buffer is
+    // null terminated and that the null is included in the length. We then
+    // subtract by 1 when constructing the underlying MemoryBuffer. This is
+    // quite sad :(
+    if (length == 0 || buffer[length - 1] != 0) {
+      return new Error("expected null terminated buffer");
+    }
+    memoryBuffer = llvm::MemoryBuffer::getMemBuffer(
+        StringRef(buffer, length - 1),
+        StringRef(bufferName, strlen(bufferName)),
+        /*RequiresNullTerminator=*/true);
```

Comment:
OOC why not just always have this be false? (I'm trying to recall, but out of cache)

Reply:
The text parsers make assumptions about being able to scan to a null, and (iirc) bad things happen when it can't satisfy this condition. Since if you get it wrong, it literally assert fails, I've gotten into the habit of holding this API with care and only using it in ways that cannot assert or read out of bounds memory -- both of which it is prone to do.